### PR TITLE
check bottle version against pkg_version

### PR DESCRIPTION
**Pull request by xu-cheng**
Addressing https://github.com/Homebrew/homebrew/pull/38587#issuecomment-92655936

### CODE REVIEW

**Comment by xu-cheng** [comment]
Don't merge it. Still WIP.

**Comment by xu-cheng** [comment]
Ping @Homebrew/owners for comments.

File: Library/Homebrew/bottle_version.rb [hunk 1]
```diff
@@ -63,3 +65,25 @@ def self._parse spec
     super
   end
 end
+
+class BottlePkgVersion < PkgVersion
```
Comment (mikemcquaid) [inline Library/Homebrew/bottle_version.rb]: Why does this need a new class? Not sure I understand.

File: Library/Homebrew/bottle_version.rb [hunk 2]
```diff
@@ -63,3 +65,25 @@ def self._parse spec
...
+  def self.parse(path)
+    path = Pathname.new(path) unless path.is_a? Pathname
+    version = BottleVersion.parse(path)
+    basename = path.basename.to_s
+    formula_name, _, revision = basename.rpartition("-#{version}")
+    if revision.match(/_(\d+)/)
+      revision = $1.to_i
```
Comment (mikemcquaid) [inline Library/Homebrew/bottle_version.rb]: If there's no match `$1.to_i` will also be `0`.

File: Library/Homebrew/cmd/bottle.rb [hunk 3]
```diff
@@ -136,7 +136,7 @@ def bottle_formula f
 
     filename = Bottle::Filename.create(f, bottle_tag, bottle_revision)
 
-    if bottle_filename_formula_name(filename).empty?
+    if bottle_filename_formula_name(filename).empty? || f.pkg_version.to_s != BottlePkgVersion.parse(filename).to_s
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/bottle.rb]: Probably worth putting this on a new line.

File: Library/Homebrew/bottle_version.rb (hunk 1 above)
Comment (xu-cheng) [inline Library/Homebrew/bottle_version.rb]: BottleVersion only parses the formula version but not revision part. So a new class to parse revision as well.

Comment (mikemcquaid) [inline Library/Homebrew/bottle_version.rb]: Why not have one class do both? The naming is kinda confusing to me.

Comment (mikemcquaid) [inline Library/Homebrew/bottle_version.rb]: Not sure I understand the subclassing either.

Comment (xu-cheng) [inline Library/Homebrew/bottle_version.rb]: > Why not have one class do both? The naming is kinda confusing to me.

It's because the subclassing things.

> Not sure I understand the subclassing either.

`BottleVersion` requires `Version` to work. `BottlePkgVersion` has the similar meaning to `PkgVersion`. At the same time, we can reuse the methods in the base class.

Comment (mikemcquaid) [inline Library/Homebrew/bottle_version.rb]: I guess it's unclear to me just from the code when I would use one over the other.

Comment (xu-cheng) [inline Library/Homebrew/bottle_version.rb]: Basically, we should use `BottlePkgVersion` only. `BottleVersion` only serves as the internal API. We do the same thing in `Version` and `PkgVersion`.

Comment (mikemcquaid) [inline Library/Homebrew/bottle_version.rb]: Seems it should maybe have a better name then and BottleVersion be folded into another class.

**Comment by jacknagel** [comment]
A subclass seems like a stretch here. This could easily be a new class with a `name` and `version` attribute, and avoid all the problems associated with subclassing.

`BottleVersion` is basically just a namespace for a class method that ultimately calls `Version.parse`, so it doesn't really need to be a subclass, either.

**Comment by xu-cheng** [comment]
@jacknagel Can you be more specific on how to write `BottleVersion` without subclass?

**Comment by xu-cheng** [comment]
Updated to make `BottleVersion` without subclass. The original `BottleVersion` was renamed to `BottleVersionParser`. And I don't know how to make it without subclass.

**Comment by mikemcquaid** [comment]
Seems fine to me but would like @jacknagel to :+1: before this gets merged.

File: Library/Homebrew/cmd/bottle.rb [hunk 4]
```diff
@@ -135,8 +135,9 @@ def bottle_formula f
     end
 
     filename = Bottle::Filename.create(f, bottle_tag, bottle_revision)
+    bottle_version = BottleVersion.parse(filename)
 
-    if bottle_filename_formula_name(filename).empty?
+    if bottle_version.formula_name.empty? || f.pkg_version != bottle_version.pkg_version
       return ofail "Add a new regex to bottle_version.rb to parse #{f.version} from #{filename}"
```
Comment (jacknagel) [inline Library/Homebrew/cmd/bottle.rb]: should this message use `f.pkg_version`?

File: Library/Homebrew/bottle_version.rb [hunk 5]
```diff
@@ -63,3 +65,28 @@ def self._parse spec
...
+  def initialize(formula_name, version, revision, bottle_revision)
+    @formula_name = formula_name
+    @version = version
+    @revision = revision
+    @bottle_revision = bottle_revision
+  end
+
+  def pkg_version
+    PkgVersion.new(@version, @revision)
+  end
```
Comment (jacknagel) [inline Library/Homebrew/bottle_version.rb]: Does it make sense to have both this and `version`? Is there a context in which knowing the version independent of the revision is useful? I don't see `version` being called anywhere, except in the test cases, where it should always be the same as `pkg_version` anyway.

Since the bottle version is always the same as the package version, maybe this should just be called `version` (in the same way it's `Keg#version` rather than `Keg#pkg_version`?

Comment (mikemcquaid) [inline Library/Homebrew/bottle_version.rb]: @xu-cheng Any thoughts on the above?

Comment (xu-cheng) [inline Library/Homebrew/bottle_version.rb]: I have unpushed commits for this PR, which includes loading formula by reading INSTALL_RECEIPTS.json,  But it needs #39524 merged first.

Comment (mikemcquaid) [inline Library/Homebrew/bottle_version.rb]: :+1:

**Comment by xu-cheng** [comment]
Updated to address previous comments. At the same time, it adds loading formula by reading `INSTALL_RECEIPT.json` feature.

File: Library/Homebrew/formulary.rb [hunk 6]
```diff
@@ -97,17 +97,31 @@ def load_file
...
-      else
-        name = name_without_version
+
+        name = full_name = @bottle_version.name
+        version = @bottle_version.version
+        tabfile = Utils.popen_read("tar", "-xOzf", @bottle_filename, "#{name}/#{version}/INSTALL_RECEIPT.json")
+        throw(:done, false) if tabfile.empty?
+
+        tap = Tab.from_tabfile_content(tabfile, @bottle_filename).tap
+        unless tap.nil? || tap == "Homebrew/homebrew"
```
Comment (mikemcquaid) [inline Library/Homebrew/formulary.rb]: Can you rework this into an `if`? I always find multi-statement `unless` really hard to follow.

File: Library/Homebrew/formulary.rb [hunk 7]
```diff
@@ -97,17 +97,31 @@ def load_file
...
+          name = full_name = bottle_name
+          throw(:done, false)
         end
-        name = bottle_name
-      else
-        name = name_without_version
+
+        name = full_name = @bottle_version.name
+        version = @bottle_version.version
+        tabfile = Utils.popen_read("tar", "-xOzf", @bottle_filename, "#{name}/#{version}/INSTALL_RECEIPT.json")
```
Comment (mikemcquaid) [inline Library/Homebrew/formulary.rb]: Given pretty much every bottle ever will have this file I wonder whether it should just replace the Bottle version detection from the filename altogether? It's a much nicer approach.

Comment (xu-cheng) [inline Library/Homebrew/formulary.rb]: So remove `BottleVersion` completely?

Comment (mikemcquaid) [inline Library/Homebrew/formulary.rb]: If we can read the version from `INSTALL_RECEIPT`: I think so, yeh. Deleting code is always good (particularly my horrible code around bottles) :wink:

Comment (xu-cheng) [inline Library/Homebrew/formulary.rb]: It's doable. We can use below code to get version and formula name.
```ruby
receipt_file = Utils.popen_read("tar", "-tzf", bottle_file, "*/*/INSTALL_RECEIPT.json").chomp
name, version, _ = receipt_file.split("/", 3)
```

However, it requires large code clean so I plan to create a new PR. Also, I will change the format of `*.bottle.rb` to include formula name if we want to remove `BottleVersion` completely

Comment (mikemcquaid) [inline Library/Homebrew/formulary.rb]: Sounds great :+1:
